core/src/main/java/org/elasticsearch/index/query/InnerHitBuilder.java

```diff
@@ -165,7 +178,35 @@ public InnerHitBuilder(StreamInput in) throws IOException {
         }
         highlightBuilder = in.readOptionalWriteable(HighlightBuilder::new);
         query = in.readNamedWriteable(QueryBuilder.class);
-        innerHitsBuilder = in.readOptionalWriteable(InnerHitsBuilder::new);
+        if (in.readBoolean()) {
+            int size = in.readVInt();
+            childInnerHits = new HashMap<>(size);
+            for (int i = 0; i < size; i++) {
+                childInnerHits.put(in.readString(), new InnerHitBuilder(in));
+            }
+        }
+    }
+
+    /**
+     * Copy constructor
+     */
+    private InnerHitBuilder(InnerHitBuilder other) {
+        name = other.name;
+        nestedPath = other.nestedPath;
+        parentChildType = other.parentChildType;
+        from = other.from;
+        size = other.size;
+        explain = other.explain;
+        version = other.version;
+        trackScores = other.trackScores;
+        fieldNames = other.fieldNames;
+        fieldDataFields = other.fieldDataFields;
+        scriptFields = other.scriptFields;
+        fetchSourceContext = other.fetchSourceContext;
+        sorts = other.sorts;
+        highlightBuilder = other.highlightBuilder;
+        query = other.query;
+        childInnerHits = other.childInnerHits;
```

Comment:
also can we have a test for this method? Something that creates a random InnerHitBuilder, makes a copy, ensure that equals returns true, then modifies the copy and makes sure equals return false?